include/datatypes/impl/chunk_element_subview.hpp

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include "enumerations/dimension.hpp"
+#include <Kokkos_Core.hpp>
+
+namespace specfem::datatype::impl {
+
+template <typename ViewType, typename PointViewType, typename IndexType>
+struct VectorChunkSubview2D {
+  ViewType &view;
+  const IndexType &index;
+
+  VectorChunkSubview2D(ViewType &view, const IndexType &index)
+      : view(view), index(index) {}
+
+  KOKKOS_INLINE_FUNCTION
+  constexpr auto &operator()(const int &icomp) {
+    return view(index.ispec, index.iz, index.ix, icomp);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  constexpr auto operator()(const int &icomp) const {
+    return view(index.ispec, index.iz, index.ix, icomp);
+  }
```

Comment:
```suggestion
  template <specfem::dimension::type D = IndexType::dimension_tag, typename std::enable_if_t<D == specfem::dimension::type::dim2, int> = 0>
  KOKKOS_INLINE_FUNCTION
  constexpr auto operator()(const int &icomp) const {
    return view(index.ispec, index.iz, index.ix, icomp);
  }
```
You'd need similar changes for other functions.

Reply:
Updated.